### DISCUSSION TITLE
Is it possible to create my own SpriteAtlasTexture?

### DISCUSSION THREAD

**Discussion by enjarai** [FabricMC/fabric-api#2563] [Mod Dev Support] [answered]
I'm trying to create my own instance of `SpriteAtlasTexture` to let me load textures completely dynamically. Since it has to be dynamic, I can't use `ClientSpriteRegistryCallback` for this. 

The problem is that when trying to instantiate `SpriteAtlasTexture`, my JVM crashes with an EXCEPTION_ACCESS_VIOLATION, even though Minecraft itself creates a bunch of them at startup as well. See attached log for full crash.

[hs_err_pid5016.log](https://github.com/FabricMC/fabric/files/9687400/hs_err_pid5016.log)

I'm completely at my wits end with this, I can't find any other mods using atlasses in this way.

**Comment by Technici4n** [comment]
It can't be entirely dynamic since the sprites are baked into a texture atlas during the resource reload.

**Reply by Technici4n** [reply]
Oh nevermind you want to create a new atlas

**Reply by enjarai** [reply]
Well by completely dynamic I mean that the specific sprites that are baked at reload are not hardcoded, and can change with every reload. They don't need to change inbetween reloads. I could've probably worded that better. 😅

**Reply by Technici4n** [reply]
The event is fired every reload.

**Reply by enjarai** [reply]
Hmmm I see, then is there a safe way to get a `ResourceManager` within that event? The specific textures that need to be loaded are specified in files inside the assets. Or is there some way I can make sure my `ResourceReloadListener` runs *before* the atlasses get made? Cause in that case I could get the values from statics there.

**Reply by Technici4n** [reply]
Based on the code it looks like it should be safe:tm: to access it via `MinecraftClient`

**Reply by enjarai** [reply]
And I should be able to specify an arbitrary id with `ClientSpriteRegistryCallback.event()` to create an atlas with that id? Or am I misunderstanding this?

**Reply by Technici4n** [reply]
No idea - you should generate the MC sources and inspect the code.

**Comment by warjort** [comment, marked as answer]
I think you are probably overthinking this? :-) 

I believe your problem from the original post is that you are crashing the game because you are trying to create a SpriteAtlasTexture before the RenderSystem (actually lwjgl) has been initialised.
It's probably a bug that it seg faults instead of throwing a java exception.

I think you just need to delay its construction until after the RenderSystem has been inited().

One way would be make your ResourceReloadListener a thin wrapper that delegates to a "lazily" constructed SpriteAtlasHolder implementation.
By the time the game is loading resource packs, the RenderSystem has been initialised.

NOTE: The resource reload processing is multi-threaded. You will need to be careful that your lazy construction (memoization) is threadsafe.

**Reply by enjarai** [reply]
Omg that was actually the problem this whole time, I was initializing it too early! It makes total sense, but I've been stumped on this for *so long*, I knew I had to have been doing *something* stupid. Tysm for this!
